pkg/runtime/serializer/protobuf/protobuf.go

```diff
@@ -0,0 +1,418 @@
+// returned size will not be correct if RawJSOn is set on unk.
+func estimateUnknownSize(unk *runtime.Unknown, byteSize uint64) uint64 {
+	size := uint64(unk.Size())
+	// protobuf uses 1 byte for the tag, a varint for the length of the array (at most 8 bytes - uint64 - here),
+	// and the size of the array.
+	size += 1 + 8 + byteSize
+	return size
+}
+
+// NewImplicitSerializer creates a Protobuf serializer that handles encoding versioned objects into the proper wire form. If typer
+// is not nil, the object has the group, version, and kind fields set. This serializer does not provide type information for the
+// encoded object, and thus is not self describing (callers must know what type is being described in order to decode).
+//
+// This encoding scheme is experimental, and is subject to change at any time.
+func NewImplicitSerializer(creater runtime.ObjectCreater, typer runtime.Typer) *ImplicitSerializer {
+	return &ImplicitSerializer{
+		creater: creater,
+		typer:   typer,
+	}
+}
+
+// ImplicitSerializer encodes and decodes objects without wrapping type info directly onto the wire.
+type ImplicitSerializer struct {
+	creater runtime.ObjectCreater
+	typer   runtime.Typer
+}
+
+// Decode attempts to convert the provided data into a protobuf message, extract the stored schema kind, apply the provided default
+// gvk, and then load that data into an object matching the desired schema kind or the provided into. If into is *runtime.Unknown,
+// the raw data will be extracted and no decoding will be performed. If into is not registered with the typer, then the object will
+// be straight decoded using normal protobuf unmarshalling (the MarshalTo interface). If into is provided and the original data is
+// not fully qualified with kind/version/group, the type of the into will be used to alter the returned gvk. On success or most
+// errors, the method will return the calculated schema kind.
+func (s *ImplicitSerializer) Decode(originalData []byte, gvk *unversioned.GroupVersionKind, into runtime.Object) (runtime.Object, *unversioned.GroupVersionKind, error) {
+	if into == nil {
+		return nil, nil, fmt.Errorf("this serializer requires an object to decode into: %#v", s)
+	}
+
+	if versioned, ok := into.(*runtime.VersionedObjects); ok {
+		into = versioned.Last()
+		obj, actual, err := s.Decode(originalData, gvk, into)
+		if err != nil {
+			return nil, actual, err
+		}
+		if into != nil && into != obj {
+			versioned.Objects = []runtime.Object{obj, into}
+		} else {
+			versioned.Objects = []runtime.Object{obj}
+		}
+		return versioned, actual, err
+	}
+
+	if len(originalData) == 0 {
+		// TODO: treat like decoding {} from JSON with defaulting
+		return nil, nil, fmt.Errorf("empty data")
+	}
+	data := originalData
+
+	actual := &unversioned.GroupVersionKind{}
+	copyKindDefaults(actual, gvk)
+
+	if intoUnknown, ok := into.(*runtime.Unknown); ok && intoUnknown != nil {
+		intoUnknown.RawJSON = data
```

Comment:
We should be renaming it to "data" or "Raw" or similar.  The proposal
discusses that but neither this nor the watch branch implements it.  We
need to add contentEncoding and contentType, although I'd like for the
proposal to get approval there.

On Thu, Mar 10, 2016 at 8:08 AM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/runtime/serializer/protobuf/protobuf.go
> <https://github.com/kubernetes/kubernetes/pull/20377#discussion_r55675490>
> :
>
> > +			versioned.Objects = []runtime.Object{obj}
> > +		}
> > +		return versioned, actual, err
> > +	}
> > +
> > +	if len(originalData) == 0 {
> > +		// TODO: treat like decoding {} from JSON with defaulting
> > +		return nil, nil, fmt.Errorf("empty data")
> > +	}
> > +	data := originalData
> > +
> > +	actual := &unversioned.GroupVersionKind{}
> > +	copyKindDefaults(actual, gvk)
> > +
> > +	if intoUnknown, ok := into.(*runtime.Unknown); ok && intoUnknown != nil {
> > +		intoUnknown.RawJSON = data
>
> I'm wondering about this one.
> Basically my doubts are about is is that "data" is not a JSON, but it's a
> serialized proto-buffer.
> So, unless I'm not missing something we:
>
>    - either need another RawProto field in runtime.Unknown
>    - or we need to rename it to RawData and add another field
>    "RawDataType" or sth like that.
>
> Or am I missing something here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20377/files#r55675490>.
>